src/main/java/com/redhat/rhjmc/containerjfr/net/web/WebServer.java

```diff
@@ -0,0 +1,389 @@
+        this.reportGenerator = reportGenerator;
+
+        if (env.getEnv(USE_LOW_MEM_PRESSURE_STREAMING_ENV) != null) {
+            logger.info("low memory pressure streaming enabled for web server");
+        } else {
+            logger.info("low memory pressure streaming disabled for web server");
+        }
+    }
+
+    private void refreshAvailableRecordings() throws FlightRecorderException {
+        recordings.clear();
+        downloadCounts.clear();
+
+        if (this.service != null) {
+            this.service.getAvailableRecordings().forEach(this::addRecording);
+        }
+    }
+
+    @Override
+    public void connectionChanged(JFRConnection connection) {
+        if (connection != null) {
+            this.service = connection.getService();
+
+            try {
+                refreshAvailableRecordings();
+            } catch (FlightRecorderException e) {
+                logger.warn(e);
+                throw new RuntimeException(e);
+            }
+        } else {
+            this.service = null;
+        }
+    }
+
+    public void start() throws FlightRecorderException, SocketException, UnknownHostException {
+        if (this.server.isAlive()) {
+            return;
+        }
+
+        refreshAvailableRecordings();
+        server.start();
+
+        Router router = Router.router(server.getVertx()); // a vertx is only available after server started
+
+        // error page handler
+        Handler<RoutingContext> failureHandler = ctx -> {
+            HttpStatusException exception;
+            if (ctx.failure() instanceof HttpStatusException) {
+                exception = (HttpStatusException) ctx.failure();
+            } else {
+                exception = new HttpStatusException(500, ctx.failure());
+            }
+
+            if (exception.getStatusCode() < 500) {
+                logger.warn(exception);
+            } else {
+                logger.error(exception);
+            }
+
+            String payload = exception.getPayload() != null ? exception.getPayload() : exception.getMessage();
+            ctx.response()
+                    .putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_PLAINTEXT)
+                    .setStatusCode(exception.getStatusCode())
+                    .setStatusMessage(exception.getMessage())
+                    .end(payload);
+        };
+
+        router.get("/clienturl").handler(this::handleClientUrlRequest).failureHandler(failureHandler);
+
+        router.get("/grafana_datasource_url").handler(this::handleGrafanaDatasourceUrlRequest)
+                .failureHandler(failureHandler);
+
+        router.get("/grafana_dashboard_url").handler(this::handleGrafanaDashboardUrlRequest)
+                .failureHandler(failureHandler);
+
+        router.get("/recordings/:name").blockingHandler(ctx -> {
+            String recordingName = ctx.pathParam("name");
+            if (recordingName != null && recordingName.indexOf(".jfr") == recordingName.length() - 4) {
```

Comment:
This looks unsafe. If I have a recording with a name of length 3, then `recordingName.indexOf(".jfr")` must be `-1` (because that substring is longer than the whole recording name), and `recordingName.length() - 4` is also `-1`. So we fall into this if-statement and attempt to take the substring of the name from 0 to -1.

Reply:
Ah, nice catch. I knew, `.indexOf()` returns `-1` if not found, but didn't realize this flaw. I'll address this problem in the next commit.

Reply:
You can probably just use `.endsWith(".jfr")`